### PR TITLE
Add QgsFeature::attributeMap() method to return feature's attributes as a field name -> value dictionary

### PR DESCRIPTION
**Pull request by nyalldawson**
The lack a matching function has long bugged me...

### CODE REVIEW

File: src/core/qgsfeature.cpp [hunk 1]
```diff
@@ -127,6 +127,17 @@ QgsAttributes QgsFeature::attributes() const
   return d->attributes;
 }
 
+QVariantMap QgsFeature::attributeMap() const
+{
+  QVariantMap res;
+  const int fieldSize = d->fields.size();
+  for ( int i = 0; i < d->attributes.size() && i < fieldSize; ++i )
```
Comment (m-kuhn) [inline src/core/qgsfeature.cpp:134]: What do you think about a sanity check here, if fieldSize != attributes.size() return an empty map and complain on all possible channels?

**Comment by m-kuhn** [comment]
Nice !

File: src/core/qgsfeature.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/core/qgsfeature.cpp:134]: I'll make it raise a ValueException!

Comment (m-kuhn) [reply src/core/qgsfeature.cpp:134]: Nice!
A check here for c++ use would still be worth it?

Comment (nyalldawson) [reply src/core/qgsfeature.cpp:134]: Yeah I'll add a debug msg for that.

Any objections to me backporting this to 3.22?

Comment (m-kuhn) [reply src/core/qgsfeature.cpp:134]: Works for me

**Comment by nyalldawson** [comment]
@m-kuhn done

**Review by lbartoletti — APPROVED** [review]

File: src/core/qgsfeature.cpp [hunk 2]
```diff
@@ -127,6 +127,23 @@ QgsAttributes QgsFeature::attributes() const
...
+{
+  QVariantMap res;
+  const int fieldSize = d->fields.size();
+  const int attributeSize = d->attributes.size();
+  if ( fieldSize != attributeSize )
+  {
+    QgsDebugMsg( QStringLiteral( "Attribute size (%1) does not match number of fields (%2)" ).arg( attributeSize ).arg( fieldSize ) );
+  }
+
+  for ( int i = 0; i < attributeSize && i < fieldSize; ++i )
```
Comment (m-kuhn) [inline src/core/qgsfeature.cpp:140]:
```suggestion
  if ( fieldSize != attributeSize )
  {
    QgsDebugMsg( QStringLiteral( "Attribute size (%1) does not match number of fields (%2)" ).arg( attributeSize ).arg( fieldSize ) );
    return QVariantMap();
  }

  for ( int i = 0; i < attributeSize; ++i )
```

That's what I meant to propose

**Review by m-kuhn — APPROVED** [review]

**Pull request merged**